File: ios/Classes/FlutterNativeEventsPlugin.swift
import Flutter
import UIKit

public class FlutterNativeEventsPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let methodChannel = FlutterMethodChannel(
      name: "flutter_native_events/methods",
      binaryMessenger: registrar.messenger()
    )
    let eventChannel = FlutterEventChannel(
      name: "flutter_native_events/events",
      binaryMessenger: registrar.messenger()
    )
    let instance = FlutterNativeEventsPlugin()
    registrar.addMethodCallDelegate(instance, channel: methodChannel)
    eventChannel.setStreamHandler(instance)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "init":
      NativeEventsBridge.shared.configure(call.arguments as? [String: Any])
      result(nil)
    case "dispose":
      NativeEventsBridge.shared.dispose()
      result(nil)
    case "emit":
      guard let arguments = call.arguments as? [String: Any] else {
        result(FlutterError(code: "invalid_arguments", message: "Arguments must be a map.", details: nil))
        return
      }
      do {
        try NativeEventsBridge.shared.handleFlutterEvent(arguments)
        result(nil)
      } catch {
        result(FlutterError(code: "emit_failed", message: error.localizedDescription, details: nil))
      }
    case "request":
      guard let arguments = call.arguments as? [String: Any] else {
        result(FlutterError(code: "invalid_arguments", message: "Arguments must be a map.", details: nil))
        return
      }
      NativeEventsBridge.shared.handleFlutterRequest(arguments, result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    NativeEventsBridge.shared.attachSink(events)
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    NativeEventsBridge.shared.attachSink(nil)
    return nil
  }
}

File: android/src/main/kotlin/com/flutter_native_events/FlutterNativeEventsPlugin.kt
package com.flutter_native_events

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class FlutterNativeEventsPlugin :
    FlutterPlugin,
    MethodChannel.MethodCallHandler,
    EventChannel.StreamHandler {
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(binding.binaryMessenger, "flutter_native_events/methods")
        eventChannel = EventChannel(binding.binaryMessenger, "flutter_native_events/events")

        methodChannel.setMethodCallHandler(this)
        eventChannel.setStreamHandler(this)
    }

    override fun onMethodCall(
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        when (call.method) {
            "init" -> {
                NativeEventsBridge.configure(call.arguments as? Map<*, *>)
                result.success(null)
            }
            "dispose" -> {
                NativeEventsBridge.dispose()
                result.success(null)
            }
            "emit" -> {
                val arguments = call.arguments as? Map<*, *>
                if (arguments == null) {
                    result.error("invalid_arguments", "Arguments must be a map.", null)
                    return
                }
                try {
                    NativeEventsBridge.handleFlutterEvent(arguments)
                    result.success(null)
                } catch (error: Throwable) {
                    result.error("emit_failed", error.message, null)
                }
            }
            "request" -> {
                val arguments = call.arguments as? Map<*, *>
                if (arguments == null) {
                    result.error("invalid_arguments", "Arguments must be a map.", null)
                    return
                }
                NativeEventsBridge.handleFlutterRequest(arguments, result)
            }
            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        NativeEventsBridge.attachSink(events)
    }

    override fun onCancel(arguments: Any?) {
        NativeEventsBridge.attachSink(null)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        NativeEventsBridge.dispose()
    }
}
